src/rpc/client.cpp

```diff
@@ -15,118 +16,245 @@
+    { "z_getmigrationstatus",        {{}, {o}} },
+    { "z_shieldcoinbase",            {{s, s}, {o, o}} },
+    { "z_mergetoaddress",            {{o, s}, {o, o, o, s}} },
+    { "z_listoperationids",          {{}, {s}} },
+    { "z_getnotescount",             {{}, {o, o}} },
+    // server
+    { "help",                        {{}, {s}} },
+    { "setlogfilter",                {{s}, {}} },
+    { "stop",                        {{}, {o}} },
 };
 
-set<int> paramsToConvertFor(const string& method) {
+std::string FormatConversionFailure(const std::string& method, const ConversionFailure& failure) {
+    return std::visit(match {
+            [&](const UnknownRPCMethod&) {
+                return tinyformat::format("Unknown RPC method, %s", method);
+            },
+            [&](const WrongNumberOfArguments& err) {
+                return tinyformat::format(
+                        "%s for method, %s. Needed between %u and %u, but received %u",
+                        err.providedArgs < err.requiredParams
+                        ? "Not enough arguments"
+                        : "Too many arguments",
+                        method,
+                        err.requiredParams,
+                        err.requiredParams + err.optionalParams,
+                        err.providedArgs);
+            },
+            [](const UnparseableArgument& err) {
+                return std::string("Error parsing JSON:") + err.unparsedArg;
+            }
+        }, failure);
+}
+
+optional<pair<vector<bool>, vector<bool>>> paramsToConvertFor(const string& method) {
     auto search = rpcCvtTable.find(method);
     if (search != rpcCvtTable.end()) {
         return search->second;
     } else {
-        return set<int>();
+        return nullopt;
     }
 }
 
-/** Non-RFC4627 JSON parser, accepts internal values (such as numbers, true, false, null)
- * as well as objects and arrays.
- */
-UniValue ParseNonRFCJSONValue(const string& strVal)
+optional<UniValue> ParseNonRFCJSONValue(const string& strVal)
 {
     UniValue jVal;
-    if (!jVal.read(string("[")+strVal+string("]")) ||
-        !jVal.isArray() || jVal.size()!=1)
-        throw runtime_error(string("Error parsing JSON:")+strVal);
-    return jVal[0];
+    if (jVal.read(string("[")+strVal+string("]")) && jVal.isArray() && jVal.size() == 1) {
+        return jVal[0];
+    } else {
+        return nullopt;
+    }
 }
 
-/** Convert strings to command-specific RPC representation */
-UniValue RPCConvertValues(const string &strMethod, const vector<string> &strParams)
+tl::expected<UniValue, ConversionFailure>
+RPCConvertValues(const string &method, const vector<string> &strArgs)
 {
     UniValue params(UniValue::VARR);
+    auto paramsToConvert = paramsToConvertFor(method);
+    vector<bool> requiredParams{};
+    vector<bool> optionalParams{};
+    if (paramsToConvert.has_value()) {
+        requiredParams = paramsToConvert.value().first;
+        optionalParams = paramsToConvert.value().second;
```

Comment:
This can be unpacked atomically:
```suggestion
        std::tie(requiredParams, optionalParams) = paramsToConvert.value();
```